### ISSUE TITLE
0 == I is not False

### ISSUE THREAD

**Issue by ptjb** [mathics/Mathics#1168]
TL;DR 
---------

0 == I  should be False, but equality comparison between Integer/Rational/Real and Complex is not possible, nor is explicitly constructing all my numbers as Complex. Mathics 2.0.1.dev.

Background
-----------------

I am doing work reducing complex matrices into linear combinations of other complex matrices using Mathics 2.0.1.dev. I am able to leverage Solve for this purpose and it's been working well about half of the time. The difficulty occurs when a solution cannot be found due to an incompatibility between a pure real from one matrix and a pure imaginary for the other.

e.g. consider the matrix equation
a|0 |=| 1|0
--|-|--|--|----
b|1| |0|1

where a is the variable of interest and we control b. The code to solve this is:

`Solve[{a == 1, 0==0, b==0, 1==1},{a}]`

When `b=0` it finds `a=1` as expected, when `b=1` it finds no solution as expected, but when `b=I` an exception is thrown:

```
System::exception:
Exception raised: can't initialize from 'dict' without generators
 
Traceback (most recent call last):
 
File "/usr/local/lib/python3.8/site-packages/mathics_django/web/views.py", line 137, in query
result = evaluation.evaluate(expr, timeout=settings.TIMEOUT)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/evaluation.py", line 352, in evaluate
result = run_with_timeout_and_stack(evaluate, timeout, self)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/evaluation.py", line 112, in run_with_timeout_and_stack
return request()
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/evaluation.py", line 323, in evaluate
self.last_eval = query.evaluate(self)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/expression.py", line 1189, in evaluate
expr, reevaluate = expr.evaluate_next(evaluation)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/expression.py", line 1317, in evaluate_next
result = rule.apply(new, evaluation, fully=False)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/rules.py", line 63, in apply
self.pattern.match(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 238, in match
self.head.match(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 137, in match
yield_func(vars, None)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 230, in yield_head
self.get_pre_choices(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 358, in get_pre_choices
yield_func(vars)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 215, in yield_choice
self.match_leaf(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 517, in match_leaf
self.get_wrappings(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 379, in get_wrappings
yield_func(items[0])
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 513, in yield_wrapping
leaf.match(match_yield, item, vars, evaluation, fully=True,
 
File "/usr/local/lib/python3.8/site-packages/mathics/builtin/patterns.py", line 787, in match
self.pattern.match(yield_func, expression, new_vars, evaluation)
 
File "/usr/local/lib/python3.8/site-packages/mathics/builtin/patterns.py", line 972, in match
yield_func(vars, None)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 503, in match_yield
self.match_leaf(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 517, in match_leaf
self.get_wrappings(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 379, in get_wrappings
yield_func(items[0])
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 513, in yield_wrapping
leaf.match(match_yield, item, vars, evaluation, fully=True,
 
File "/usr/local/lib/python3.8/site-packages/mathics/builtin/patterns.py", line 787, in match
self.pattern.match(yield_func, expression, new_vars, evaluation)
 
File "/usr/local/lib/python3.8/site-packages/mathics/builtin/patterns.py", line 972, in match
yield_func(vars, None)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 510, in match_yield
yield_func(new_vars, items_rest)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/pattern.py", line 497, in leaf_yield
yield_func(
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/rules.py", line 40, in yield_match
new_expression = self.do_replace(expression, vars, options, evaluation)
 
File "/usr/local/lib/python3.8/site-packages/mathics/core/rules.py", line 129, in do_replace
return self.function(evaluation=evaluation, **vars_noctx)
 
File "/usr/local/lib/python3.8/site-packages/mathics/builtin/calculus.py", line 814, in apply
result = sympy.solve(sympy_eqs, vars_sympy)
 
File "/usr/local/lib/python3.8/site-packages/sympy/solvers/solvers.py", line 1099, in solve
solution = _solve_system(f, symbols, **flags)
 
File "/usr/local/lib/python3.8/site-packages/sympy/solvers/solvers.py", line 1774, in _solve_system
subsol = _solve_system(subexpr, subsyms, **flags)
 
File "/usr/local/lib/python3.8/site-packages/sympy/solvers/solvers.py", line 1805, in _solve_system
poly = g.as_poly(*symbols, extension=True)
 
File "/usr/local/lib/python3.8/site-packages/sympy/core/expr.py", line 1109, in as_poly
poly = Poly(self, *gens, **args)
 
File "/usr/local/lib/python3.8/site-packages/sympy/polys/polytools.py", line 162, in __new__
return cls._from_expr(rep, opt)
 
File "/usr/local/lib/python3.8/site-packages/sympy/polys/polytools.py", line 292, in _from_expr
return cls._from_dict(rep, opt)
 
File "/usr/local/lib/python3.8/site-packages/sympy/polys/polytools.py", line 229, in _from_dict
raise GeneratorsNeeded(
 
sympy.polys.polyerrors.GeneratorsNeeded: can't initialize from 'dict' without generators
```
one can easily verify that in each case we have `0==0` evaluating to `True`, `0==1` evaluating to `False` but `0==I` not evaluating at all. In fact `BooleanQ[0==I]` equals `False`.

Solve can clearly handle my way of using it when all the `constant == constant` expressions are Boolean, but fails to when it finds an irreducible one, like `0==I`. I wouldn't blame Solve for this as I think that's perfectly reasonable. I think the issue here is that judging equality between an Integer/Rational/Real and a Complex is not possible. I could accept having to explicitly construct all of my numbers as Complex but `Complex[x,0]` for any `x` is never Complex, so this also fails.

Any and all suggestions are welcome.

**Comment by rocky** [comment]
Seems reasonable to allow comparison of Complex with a Rational or Real. And is a bug that we don't currently. 

Want to try to fix the code?

**Comment by mmatera** [comment]
The question is if the bug is at the level of mathics or is inside simpy...

**Comment by rocky** [comment]
I think it is in Mathics:

```
Out[1]= I == 0 # empty "Out" below means result is Python None

In[2]:= I == 2 I
Out[2]= False

```

**Comment by ptjb** [comment]
I'll have a look into it and report back!

**Comment by rocky** [comment]
https://github.com/mathics/Mathics/tree/ExactNumber-comparisions has some fixes. It seems to break other tests though.
